### ISSUE TITLE
ARM specific optimizations

### ISSUE THREAD

**Issue by Adenilson** [zlib-ng/zlib-ng#82]
Libpng has both intrinsics and hand written ASM code for ARM (on the pre-filters).

Would this zlib fork be open to contributions of a few core/hot functions targeting ARM?

**Comment by Dead2** [comment]
We would be glad to have arch-specific optimizations. Have a look at how that is currently solved for the x86-specific code in our fork, we try to have the code separated out of the main code as much as possible. Also, we prefer compiler intrinsics if possible.

Also, have a look at for example this commit
https://github.com/mtl1979/zlib-ng/commit/5ddec4b5244fc455b9e689c2ddd733c1ad872ca1
That code has not yet been pulled into zlib-ng, but @mtl1979 has been preparing some arm-specific optimizations for merging, so you should remain compatible with that.

**Comment by Adenilson** [comment]
That is great to hear!

To help sync the patches, when @mtl1979 is planning to land his ARM specific changes?

**Comment by mtl1979** [comment]
@Adenilson We're going to land my ARM specific changes after we have tested them on real hardware to see if there is some speed tuning that still can be done or some hidden bugs I have not noticed.

**Comment by Adenilson** [comment]
@mtl1979 that is pretty cool, let me know if is there anything I could do to help.
:-)

@Dead2 
Chromium uses a forked zlib (with a few of the Intel patches, but not all of them) and there are discussions about migrating to a 'new' zlib and zlib-ng is one of the good candidates.

I recently implemented a NEON-ized version of Adler32 checksum (https://codereview.chromium.org/2676493007/, about 3x faster) and I'm looking forward to upstream this patches to zlib-ng instead of forking even more the zlib used in Chromium.

One question: what is the sync frequency of zlib-ng with the canonical zlib?

**Comment by mtl1979** [comment]
@Adenilson The patch could be added to zlib-ng with minor modifications, but zlib-ng tries to avoid K&R workarounds as we target only modern compilers... That limitation makes our code cleaner and easier to maintain. We also have to remember that longer we keep variables around, more likely gcc will run out of registers when generating assembler code from C or C++ code... Even though ARM has 31 general purpose registers, gcc only uses subset of them for user code. When I write code, I check the gcc intermediate output to see if gcc has run out of registers and adjust the code.

**Comment by mtl1979** [comment]
@Dead2 If we implement NEON optimization for adler32 on ARM/AARCH64, should we also implement same optimization using SSE/AVX on Intel?

**Comment by Dead2** [comment]
@Adenilson We try to keep up with upstream zlib, but lately work and real life has matched perfectly with Mark Adler's unexpected much increased activity, so we have been backlogged for a long time. I spent every free moment last week playing catch-up, and I made some progress yesterday as well. You can see the current progress in the hacknslash5 branch, it currently does not pass travis checks though since I pushed my last progress untested before leaving work yesterday. But I am synced up to 31 Jan 2016 now, so things are looking better.

Porting some of the upstream refactoring commits has been challenging, because they sometimes subtly break things in a completely different part of the code due to our massive changes.

We would be glad to have more people do active testing and/or contribute to the project. One thing we really need is unit tests, because testing every corner of this codebase is really hard because zlib exports so many different methods of doing compression/decompression. I don't think I will be requiring new code to have unit tests (yet), but I sure would like to. :)

@mtl1979 I was thinking the same thing, it would be nice to look into that possibility for sure. It would likely also benefit decompression as well as compression, and speeding up decompression has been something we haven't been doing a lot of so far.

**Comment by mtl1979** [comment]
@Dead2 @Adenilson I will try to get the neon-optimized adler32 integrated to my arm-test branch... It might take few hours before I can push it to GitHub, but after that, it's open for comments and proofreading ;)

**Comment by Adenilson** [comment]
@mtl1979 That is awesome you can integrate the NEON adler32 in your ARM branch!

Let me know if is there any pending issue that I could help.

Concerning register use, I agree that is advisable to avoid misusing them, and I will welcome any change that makes the code even faster.
:-)

**Comment by Adenilson** [comment]
@Dead2 You touched a key issue that I feel Opensource could do better: unit tests.

While developing this code, I wrote quite a few unit tests and compared the results calculated by the NEON adler X zlib adler32. That helped a lot to spot bugs and stop them from hitting production.

I feel it would be helpful to maybe upstream this tests, otherwise there is no other proper way to ensure correctness.

Specially with platform specific code, I believe that any new implementation must come with a test suite and should only be accepted if the results match the canonical C implementation.

On Chromium/Blink, Google test (gtest) is used to write unit tests and it works fine on Android/Linux/etc. It also integrates nicely with CMake.

An example of this test strategy that I'm suggesting can be found on this patch I contributed to Chromium last month:
https://cs.chromium.org/chromium/src/third_party/WebKit/Source/platform/image-encoders/JPEGImageEncoderTest.cpp

Basically the test compares the results of converting RGBAtoRGB, testing scalar version X the vectorial implementation. Besides testing for correctness, I also wrote a test for performance on that patch: the walltime for execution of the NEON-ized version is expected to be smaller than the scalar version (and the test should fail if it is not).

Does zlib-ng support unit tests?

Finally, it would be nice to validate zlib-ng with libpng, as they are both associated projects (e.g. does pngtest works fine with zlib-ng?). While working on this optimization for Adler, I found one case where the zlib test apps would work fine but pngtest would fail with a specific image.
